spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,301 @@
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), -1, true);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    this.pendingTasks = getChangesWithRateLimit(start.snapshotId(), start.index(),
+        start.isStartingSnapshotId(), maxFilesPerBatch);
+    IndexedTask last = Iterables.isEmpty(pendingTasks) ? null : Iterables.getLast(pendingTasks);
+
+    if (last == null) {
+      return start;
+    } else {
+      boolean isStarting = last.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(last.snapshotId(), last.index(), isStarting);
+    }
+  }
+
+  @VisibleForTesting
+  CloseableIterable<IndexedTask> getChangesWithRateLimit(long snapshotId, int index, boolean isStarting, int maxFiles) {
+    List<CloseableIterable<IndexedTask>> snapshotTasks = Lists.newArrayList();
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+    int currentFiles = 0;
+
+    if (isStarting) {
+      CloseableIterable<FileScanTask> iter = buildTableScan()
+          .option("use-worker-pool", "false")
+          .useSnapshot(snapshotId)
+          .planFiles();
+      CloseableIterable<IndexedTask> tasks = filterTasks(indexTasks(iter, snapshotId), snapshotId, index);
+      Pair<Integer, CloseableIterable<IndexedTask>> limitedTasks = limitTasks(tasks, currentFiles, maxFiles);
+      currentFiles = limitedTasks.getLeft();
+      snapshotTasks.add(limitedTasks.getRight());
+
+      if (currentFiles >= maxFiles) {
+        // If the files to be processed already reach to max-files-batch, will not continue to plan files.
+        return CloseableIterable.concat(snapshotTasks);
+      }
+    }
+
+    Long start = isStarting ? Long.valueOf(snapshotId) : table.snapshot(snapshotId).parentId();
+    Preconditions.checkState(start != null, "Start snapshot id should exist");
+    ImmutableList<Long> snapshotIds = ImmutableList.<Long>builder()
+        .addAll(SnapshotUtil.snapshotIdsBetween(table, start, currentSnapshotId))
+        .add(start)
+        .build()
+        .reverse();
+
+    for (int i = 0; i < snapshotIds.size() - 1; i++) {
+      long fromSnapshotId = snapshotIds.get(i);
+      long toSnapshotId = snapshotIds.get(i + 1);
+
+      CloseableIterable<FileScanTask> iter = buildTableScan()
+          .option("use-worker-pool", "false")
+          .appendsBetween(fromSnapshotId, toSnapshotId)
+          .planFiles();
+      CloseableIterable<IndexedTask> tasks = filterTasks(indexTasks(iter, toSnapshotId), snapshotId, index);
+      Pair<Integer, CloseableIterable<IndexedTask>> limitedTasks = limitTasks(tasks, currentFiles, maxFiles);
+      currentFiles = limitedTasks.getLeft();
```

Comment:
I think this should be `+=`, right?

Reply:
No, the returned value of `limitTasks` is already a cumulated value, so don't need to add again.